### PR TITLE
Feat/manual peering

### PR DESCRIPTION
**Pull request by georgysavva**
# Description of change

This PR introduces the separation of neighbors in the gossip layer by groups. Right now we have two groups: "auto" (the existing type of neighbors that come from the autopeering) and "manual" (the new type of neighbors that come from a config file or HTTP API). The gossip manager still has a single list of all neighbors, but now every neighbor holds information about to which group it belongs. When autopeering tries to drop a neighbor it will be able to drop only neighbors in the "auto" group and won't be able to drop "manual" neighbors. It also works vice versa (manual call can't drop "auto" neighbors).

This PR also adds a way to pass neighbors to the gossip manager via JSON configuration file and JSON HTTP API.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement 

## How the change has been tested

For now, I tested only manually, but we are going to use those changes in integration tests and it will be covered.

Depends on https://github.com/iotaledger/hive.go/pull/253.

### CODE REVIEW

**Review by capossele — APPROVED** [review]
looks monumental to me

File: plugins/gossip/parameters.go [hunk 1]
```diff
@@ -13,10 +13,15 @@ const (
 	CfgGossipAgeThreshold = "gossip.ageThreshold"
 	// CfgGossipTipsBroadcastInterval the interval in which the oldest known tip is re-broadcast.
 	CfgGossipTipsBroadcastInterval = "gossip.tipsBroadcaster.interval"
+	// CfgGossipDisableAutopeering if set to true the autopeering layer won't manage neighbors in the gossip layer.
```
Comment (jonastheis) [inline plugins/gossip/parameters.go:16]: All these and the "old" parameters can (and maybe should) be simplified using the new framework like here: https://github.com/iotaledger/goshimmer/blob/3f56dd4a2ccf8f086e188f07c6f23ecb741824be/plugins/messagelayer/parameters.go#L10

File: plugins/gossip/webapi.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	"golang.org/x/xerrors"
+
+	"github.com/iotaledger/goshimmer/packages/gossip"
+	"github.com/iotaledger/goshimmer/plugins/webapi"
+	"github.com/iotaledger/goshimmer/plugins/webapi/jsonmodels"
+)
+
+func configureWebAPI() {
+	webapi.Server().POST("gossip/neighbors", addNeighborsHandler)
+	webapi.Server().DELETE("gossip/neighbors", dropNeighborsHandler)
```
Comment (lzpap) [inline plugins/gossip/webapi.go:19]: should we add a GET route to get the neighbors from gossip? we can fetch the autopeering neighbors from `autopeering/neighbors`, but I guess those will not include manual peers

File: plugins/gossip/parameters.go (hunk 1 above)
Comment (georgysavva) [reply plugins/gossip/parameters.go:16]: Okay I will look into this

File: plugins/gossip/webapi.go (hunk 2 above)
Comment (georgysavva) [reply plugins/gossip/webapi.go:19]: Yeah, I think it's good to add such API

File: plugins/gossip/parameters.go (hunk 1 above)
Comment (georgysavva) [reply plugins/gossip/parameters.go:16]: Sorry, I don't have time to do it in the scope of this PR. Let's work on this later.

**Review by jonastheis — CHANGES_REQUESTED** [review]
I think it looks quite nice! Great work :) 

Though I have some questions that might be addressed in a separate PR I suppose:
- display neighbor types differently in dashboard? e.g. even if not connected and in different grouping than autopeering neighbors?
- how to display in (auto)peering visualizer? is it automatically working?

I'm not entirely clear on how the connection is actually established. Do the peers need to be synchronized in calling `m.gm.AddInbound` and `m.gm.AddOutbound`? Or how long does the  gossip wait to accept a connection in `m.gm.AddInbound`? If it waits, would that block further execution in `syncGossipNeighbors`? 

Or should a `connDirectionInbound` node always be ready to accept a connection?

File: plugins/manualpeering/webapi.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		)
+	}
+	if err := removePeers(peersToRemove); err != nil {
+		log().Errorw(
+			"Can't remove some of the peers from the HTTP request",
+			"err", err,
+		)
+		return c.JSON(http.StatusInternalServerError, jsonmodels.NewErrorResponse(err))
+	}
+	return c.JSON(http.StatusOK, map[string]bool{"ok": true})
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply plugins/manualpeering/webapi.go:55]: Ok

File: plugins/manualpeering/webapi.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/iotaledger/goshimmer/plugins/webapi/jsonmodels"
+)
+
+func configureWebAPI() {
+	webapi.Server().POST("manualpeering/knownPeers", addPeersHandler)
+	webapi.Server().DELETE("manualpeering/knownPeers", removePeersHandler)
+	webapi.Server().GET("manualpeering/knownPeers", getPeersHandler)
+}
+
+func addPeersHandler(c echo.Context) error {
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply plugins/manualpeering/webapi.go:22]: Yeah, we can put it there for now.

File: plugins/manualpeering/plugin.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+func Plugin() *node.Plugin {
+	pluginOnce.Do(func() {
+		plugin = node.NewPlugin(PluginName, node.Enabled, configurePlugin, runPlugin)
+	})
+	return plugin
+}
+
+func log() *logger.Logger {
+	return logger.NewLogger(PluginName)
+}
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply plugins/manualpeering/plugin.go:45]: Ok, thanks

File: plugins/manualpeering/plugin.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+// Manager is a singleton for manualpeering Manager.
+func Manager() *manualpeering.Manager {
+	managerOnce.Do(createManager)
+	return manager
+}
+
+func createManager() {
+	lPeer := local.GetInstance()
+	manager = manualpeering.NewManager(gossip.Manager(), lPeer, log())
+}
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply plugins/manualpeering/plugin.go:56]: Okay.

File: plugins/manualpeering/parameters.go [hunk 7]
```diff
@@ -0,0 +1,4 @@
+package manualpeering
+
+// CfgManualpeeringKnownPeers list of peers that will be used as known peers in the manualpeering.
+const CfgManualpeeringKnownPeers = "manualpeering.knownPeers"
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply plugins/manualpeering/parameters.go:4]: No, there is no way right now, it was too complicated and I skipped that.

File: plugins/gossip/plugin.go [hunk 8]
```diff
@@ -111,13 +115,13 @@ func configureLogging() {
...
+	mgr.NeighborsEvents(gossip.NeighborsGroupAuto).ConnectionFailed.Attach(events.NewClosure(func(p *peer.Peer, err error) {
 		log.Infof("Connection to neighbor %s / %s failed: %s", gossip.GetAddress(p), p.ID(), err)
 	}))
-	mgr.Events().NeighborAdded.Attach(events.NewClosure(func(n *gossip.Neighbor) {
+	mgr.NeighborsEvents(gossip.NeighborsGroupAuto).NeighborAdded.Attach(events.NewClosure(func(n *gossip.Neighbor) {
 		log.Infof("Neighbor added: %s / %s", gossip.GetAddress(n.Peer), n.ID())
 	}))
-	mgr.Events().NeighborRemoved.Attach(events.NewClosure(func(n *gossip.Neighbor) {
+	mgr.NeighborsEvents(gossip.NeighborsGroupAuto).NeighborRemoved.Attach(events.NewClosure(func(n *gossip.Neighbor) {
 		log.Infof("Neighbor removed: %s / %s", gossip.GetAddress(n.Peer), n.ID())
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply plugins/gossip/plugin.go:125]: I think you are right. @capossele what do you think?

File: packages/manualpeering/manualpeering.go [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+// If a new peer is added to known peers, manager will forward it to gossip and make sure it establishes a connection.
+// And vice versa, if a peer is being removed from the list of known peers,
+// manager will make sure gossip drops that connection.
+// Manager also subscribes to the gossip events and in case the connection with a manual peer fails it will reconnect.
+type Manager struct {
+	gm                      *gossip.Manager
+	log                     *logger.Logger
+	local                   *peer.Local
+	startOnce               sync.Once
+	isStarted               int32
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply packages/manualpeering/manualpeering.go:33]: Sure.

File: packages/manualpeering/manualpeering.go [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+	}
+	m.mutateKnownPeers(func(knownPeers map[identity.ID]*peer.Peer) {
+		for _, id := range ids {
+			delete(knownPeers, id)
+		}
+	})
+}
+
+// AddPeer adds a single peer to the list on known peers.
+func (m *Manager) AddPeer(p *peer.Peer) {
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply packages/manualpeering/manualpeering.go:83]: Yeah, good idea.

File: packages/manualpeering/manualpeering.go [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+	})
+}
+
+// Stop terminates internal background workers. Calling multiple times has no effect.
+func (m *Manager) Stop() error {
+	if atomic.LoadInt32(&m.isStarted) != 1 {
+		return errors.New("can't stop the manager: it hasn't been started yet")
+	}
+	m.stopOnce.Do(func() {
+		m.gm.NeighborsEvents(gossip.NeighborsGroupManual).NeighborRemoved.Detach(events.NewClosure(m.onGossipNeighborRemoved))
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply packages/manualpeering/manualpeering.go:137]: I see, thanks

**Comment by georgysavva** [comment]
> I think it looks quite nice! Great work :)
> 
> Though I have some questions that might be addressed in a separate PR I suppose:
> 
> * display neighbor types differently in dashboard? e.g. even if not connected and in different grouping than autopeering neighbors?
> * how to display in (auto)peering visualizer? is it automatically working?
> 
> I'm not entirely clear on how the connection is actually established. Do the peers need to be synchronized in calling `m.gm.AddInbound` and `m.gm.AddOutbound`? Or how long does the gossip wait to accept a connection in `m.gm.AddInbound`? If it waits, would that block further execution in `syncGossipNeighbors`?
> 
> Or should a `connDirectionInbound` node always be ready to accept a connection?

@jonastheis answering your general questions:
1. I don't quite understand what it is but sounds like it should be easy to implement.
2. I don't think it will work automatically and I am not sure whether it should be there, @capossele.
3. Ideally those calls should happen within a short time interval to establish the connection quickly. Right now the gossip blocks on `AddInbound()` call without a timeout, so it means if there is no dialing on the other side `AddInbound()` is blocked, and hence the `syncGossipNeighbors()` is blocked as well. But it's easy to fix and I will do it in the scope of this PR. I am going to add a timeout to the `AddInbound()` call and make the `syncGossipNeighbors()` process every connection in parallel.

**Review by capossele — APPROVED** [review]

File: plugins/manualpeering/webapi.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+		)
+	}
+	if err := removePeers(peersToRemove); err != nil {
+		plugin.Logger().Errorw(
+			"Can't remove some of the peers from the HTTP request",
+			"err", err,
+		)
+		return c.JSON(http.StatusInternalServerError, jsonmodels.NewErrorResponse(err))
+	}
+	return c.JSON(http.StatusNoContent, map[string]bool{"ok": true})
```
Comment (jonastheis) [inline plugins/manualpeering/webapi.go:82]:
```suggestion
	return c.NoContent(http.StatusNoContent)
```

File: plugins/manualpeering/webapi.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+	var peers []*peer.Peer
+	if err := webapi.ParseJSONRequest(c, &peers); err != nil {
+		plugin.Logger().Errorw("Failed to parse peers from the request", "err", err)
+		return c.JSON(
+			http.StatusBadRequest,
+			jsonmodels.NewErrorResponse(errors.Wrap(err, "Invalid add peers request")),
+		)
+	}
+	Manager().AddPeer(peers...)
+	return c.JSON(http.StatusOK, map[string]bool{"ok": true})
```
Comment (jonastheis) [inline plugins/manualpeering/webapi.go:51]:
```suggestion
	return c.NoContent(http.StatusNoContent)
```

**Review by jonastheis — APPROVED** [review]

**Pull request merged**